flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -804,199 +792,15 @@ public boolean restoreLatestCheckpointedState(
-
-						if (partitionableState != null) {
-							for (int j = 0; j < partitionableState.getLength(); ++j) {
-								OperatorStateHandle opParalleState = partitionableState.get(j);
-								if (opParalleState != null) {
-									List<OperatorStateHandle> opParallelStates =
-											chainParallelStates[j];
-									if (opParallelStates == null) {
-										opParallelStates = new ArrayList<>();
-										chainParallelStates[j] = opParallelStates;
-									}
-									opParallelStates.add(opParalleState);
-								}
-							}
-						}
-					}
-
-					// operator chain index -> lists with collected states (one collection for each parallel subtasks)
-					@SuppressWarnings("unchecked")
-					List<Collection<OperatorStateHandle>>[] redistributedParallelStates =
-							new List[taskState.getChainLength()];
-
-					//TODO here we can employ different redistribution strategies for state, e.g. union state. For now we only offer round robin as the default.
-					OperatorStateRepartitioner repartitioner = RoundRobinOperatorStateRepartitioner.INSTANCE;
-
-					for (int i = 0; i < chainParallelStates.length; ++i) {
-						List<OperatorStateHandle> chainOpParallelStates = chainParallelStates[i];
-						if (chainOpParallelStates != null) {
-							//We only redistribute if the parallelism of the operator changed from previous executions
-							if (parallelismChanged) {
-								redistributedParallelStates[i] = repartitioner.repartitionState(
-										chainOpParallelStates,
-										newParallelism);
-							} else {
-								List<Collection<OperatorStateHandle>> repacking = new ArrayList<>(newParallelism);
-								for (OperatorStateHandle operatorStateHandle : chainOpParallelStates) {
-									repacking.add(Collections.singletonList(operatorStateHandle));
-								}
-								redistributedParallelStates[i] = repacking;
-							}
-						}
-					}
-
-					int counter = 0;
-
-					for (int i = 0; i < newParallelism; ++i) {
-
-						// non-partitioned state
-						ChainedStateHandle<StreamStateHandle> state = null;
-
-						if (hasNonPartitionedState) {
-							SubtaskState subtaskState = taskState.getState(i);
-
-							if (subtaskState != null) {
-								// count the number of executions for which we set a state
-								++counter;
-								state = subtaskState.getChainedStateHandle();
-							}
-						}
-
-						// partitionable state
-						@SuppressWarnings("unchecked")
-						Collection<OperatorStateHandle>[] ia = new Collection[taskState.getChainLength()];
-						List<Collection<OperatorStateHandle>> subTaskPartitionableState = Arrays.asList(ia);
-
-						for (int j = 0; j < redistributedParallelStates.length; ++j) {
-							List<Collection<OperatorStateHandle>> redistributedParallelState =
-									redistributedParallelStates[j];
-
-							if (redistributedParallelState != null) {
-								subTaskPartitionableState.set(j, redistributedParallelState.get(i));
-							}
-						}
+			StateAssignmentOperation stateAssignmentOperation =
```

Comment:
I was not sure whether this class will have state or not. I might still want to break assignStates() down into smaller methods, which might lead me to introducing some state to make. Last but not least, having things more fine grained and in a class allows better unit testing and mocking, while mocking static methods is impossible. This is one of the places where we might want something like this in the future.

Not sure about the allOrNothingState, but that could be a separate change.